### PR TITLE
PLATFORM-8040 Upgrade Reverb to work with MW 1.39

### PR DESCRIPTION
**Pull request by mmeller-wikia**
https://fandom.atlassian.net/browse/PLATFORM-8040

I suggest reviewing it commit by commit as multiple files has been moved around

Depends on https://github.com/Wikia/reverb-ext/pull/18

### CODE REVIEW

File: .github/workflows/php-pr-checks.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      run: composer phpcs
+
+    - name: Run Phan static analysis
+      run: composer phan
+
+    - name: Run PHPUnit tests with coverage
+      run: composer phpunit
+
+    - name: Run Jasmine tests
+      run: npm test
```
Comment (mmeller-wikia) [inline .github/workflows/php-pr-checks.yml:58]: It's basically `assertTrue` test, but let's leave integration with Jasmine ready
https://github.com/Wikia/reverb-ext/blob/230f1da0269cdc2abf30aaa9f27cadfcca7f1769/spec/jasmine_examples/ExampleSpec.js#L1-L8

Comment (mszabo-wikia) [reply .github/workflows/php-pr-checks.yml:58]: Let's reflect the fact we're running JS tests in the check name too

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "Reverb",
-  "version": "0.0.1",
+  "version": "0.0.2",
```
Comment (mmeller-wikia) [inline package.json:3]: Bump version to invalidate pipeline cache

File: src/ReverbHookHandler.php [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+				'tooltips' => $tooltips,
+			];
+		}
+		foreach ( $preferences as $index => $preference ) {
+			if ( isset( $preference[ 'section' ] ) && $preference[ 'section' ] === 'personal/email' ) {
+				$preferences[ $index ][ 'section' ] = 'reverb/reverb-email-options';
+			}
+
+			// Reverb supercedes Fandom email preferences, so don't show them in Special:Preferences
+			// Note: this depends on Reverb being loaded after fandom extensions
```
[the remark this answers is not in the document]
Comment (mmeller-wikia) [reply src/ReverbHookHandler.php:135]: It is still needed -> Gamepedia uses alternative `Special:Preference` options for emails
image

**Review by mszabo-wikia — APPROVED** [review]

File: .github/workflows/php-pr-checks.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    - name: Cache npm packages
+      id: npm-cache
+      uses: actions/cache@v3
+      with:
+        path: vendor
+        key: ${{ runner.os }}-npm-v3-${{ hashFiles('**/package-lock.json') }}
+
+    - name: Install npm dependencies
+      if: steps.npm-cache.outputs.cache-hit != 'true'
+      run: npm install
```
[the remark this answers is not in the document]
Comment (mszabo-wikia) [reply .github/workflows/php-pr-checks.yml:46]: Yes, `ci` removes the `node_modules` folder so we shouldn't cache that but rather the npm cache dir: https://docs.npmjs.com/cli/v9/commands/npm-ci

File: src/Notification/NotificationFactory.php [hunk 5]
```diff
@@ -6,6 +6,7 @@
 use Fandom\Includes\Util\UrlUtilityService;
 use Fandom\WikiConfig\WikiVariablesDataService;
 use MediaWiki\User\UserFactory;
+use Reverb\Identifier\IdentifierService;
 use WikiDomain\WikiConfigDataService;
 
 class NotificationFactory {
```
Comment (mszabo-wikia) [inline src/Notification/NotificationFactory.php:12]: Is there a test for this class?

File: extension.json [hunk 6]
```diff
@@ -104,7 +104,8 @@
 				"DBLoadBalancer",
 				"WatchedItemStore",
 				"Fandom\\Includes\\User\\UserInfo",
-				"Fandom\\Includes\\Util\\UrlUtilityService"
+				"Fandom\\Includes\\Util\\UrlUtilityService",
+				"Reverb\\Notification\\NotificationBroadcastFactory"
```
Comment (mmeller-wikia) [inline extension.json:108]: Note: This issue was unexpected for me - when HookHandler class cannot be created, it fails silently (at least in this case)
```
ArgumentCountError from line 56 of /extensions/Reverb/src/ReverbBroadcastHookHandler.php: Too few arguments to function Reverb\ReverbBroadcastHookHandler::__construct(), 11 passed in /vendor/wikimedia/object-factory/src/ObjectFactory.php on line 247 and exactly 12 expected
	#0 /vendor/wikimedia/object-factory/src/ObjectFactory.php(247): Reverb\ReverbBroadcastHookHandler->__construct(GlobalVarConfig, MediaWiki\User\UserFactory, MediaWiki\Revision\RevisionStore, NamespaceInfo, MediaWiki\Revision\RevisionStore, MediaWiki\User\UserOptionsManager, MediaWiki\Languages\LanguageFactory, Fandom\Includes\Database\FandomLoadBalancer, Fandom\Includes\Watchlist\WatchedItemStoreOverrides, Fandom\Includes\User\UserInfo, Fandom\Includes\Util\UrlUtilityService)
	#1 /vendor/wikimedia/object-factory/src/ObjectFactory.php(152): Wikimedia\ObjectFactory\ObjectFactory::getObjectFromSpec(array, array)
	#2 /includes/HookContainer/HookContainer.php(473): Wikimedia\ObjectFactory\ObjectFactory->createObject(array)

File: src/Notification/NotificationFactory.php (hunk 5 above)
Comment (mmeller-wikia) [reply src/Notification/NotificationFactory.php:12]: No, in this extension there is only one test class (for `IdentifierService`)

**Review by mszabo-wikia — APPROVED** [review]

**Pull request merged**